language/move-vm/runtime/src/data_cache.rs

```diff
@@ -67,149 +90,164 @@ impl<'txn> TransactionDataCache<'txn> {
-                            &ty.address,
-                            &ty.module.as_str(),
-                            &ty.name.as_str(),
-                            ap
-                        )),
-                    );
+                    return Err(VMStatus::new(StatusCode::MISSING_DATA)
+                        .with_message(format!("Cannot find resource of type {}", ty_tag)))
                 }
             };
+            let gv = GlobalValue::new(val)?;
+            account_cache.data_map.insert(ty.clone(), Some(gv));
         }
-        Ok(self.data_map.get_mut(ap).expect("data must exist"))
+
+        Ok(account_cache.data_map.get_mut(ty).unwrap())
     }
 }
 
 // `DataStore` implementation for the `TransactionDataCache`
-impl<'a> DataStore for TransactionDataCache<'a> {
-    fn publish_resource(
-        &mut self,
-        ap: &AccessPath,
-        g: (FatStructType, GlobalValue),
-    ) -> VMResult<()> {
-        self.data_map.insert(ap.clone(), Some(g));
+impl<'r, 'l, C: RemoteCache> DataStore for TransactionDataCache<'r, 'l, C> {
+    fn publish_resource(&mut self, addr: AccountAddress, ty: Type, g: GlobalValue) -> VMResult<()> {
+        let account_cache = Self::get_mut_or_insert_with(&mut self.account_map, &addr, || {
+            (addr.clone(), AccountDataCache::new())
+        });
+
+        // TODO: shall we check if resource already exists here?
+        account_cache.data_map.insert(ty, Some(g));
+
         Ok(())
     }
 
     fn borrow_resource(
         &mut self,
-        ap: &AccessPath,
-        ty: &FatStructType,
+        addr: AccountAddress,
+        ty: &Type,
     ) -> VMResult<Option<&GlobalValue>> {
-        let map_entry = self.load_data(ap, ty)?;
-        Ok(map_entry.as_ref().map(|(_, g)| g))
+        Ok(self.load_data(addr, ty)?.as_ref())
     }
 
     fn move_resource_from(
         &mut self,
-        ap: &AccessPath,
-        ty: &FatStructType,
+        addr: AccountAddress,
+        ty: &Type,
     ) -> VMResult<Option<GlobalValue>> {
-        let map_entry = self.load_data(ap, ty)?;
         // .take() means that the entry is removed from the data map -- this marks the
         // access path for deletion.
-        Ok(map_entry.take().map(|(_, g)| g))
+        Ok(self.load_data(addr, ty)?.take())
     }
 
-    fn load_module(&self, module: &ModuleId) -> VMResult<Vec<u8>> {
-        match self.module_map.get(module) {
-            Some(bytes) => Ok(bytes.clone()),
-            None => {
-                let ap = AccessPath::from(module);
-                self.data_cache.get(&ap).and_then(|data| {
-                    data.ok_or_else(|| {
-                        VMStatus::new(StatusCode::LINKER_ERROR)
-                            .with_message(format!("Cannot find {:?} in data cache", module))
-                    })
-                })
+    // REVIEW: The txn data cache isn't really caching the modules read from storage. Is this really desired?
```

Comment:
I think that is an ok thing to do given the code cache will. Effectively `load_module` should only be called by the loader